// cxs/dist/sheet.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var isBrowser = typeof document !== 'undefined';

function Stylesheet() {
  var _this = this;

  this.tag = null;
  this.sheet = null;
  this.injected = false;

  this.inject = function () {
    if (_this.injected) return;

    if (isBrowser) {
      _this.tag = createStyleTag();
      _this.sheet = _this.tag.sheet;
    } else {
      _this.sheet = {
        cssRules: [],
        insertRule: function insertRule(rule) {
          _this.sheet.cssRules.push({
            cssText: rule
          });
        }
      };
    }
    _this.injected = true;
  };

  this.insert = function (rule) {
    _this.sheet.insertRule(rule, _this.sheet.cssRules.length);
  };

  this.flush = function () {
    if (!isBrowser) {
      _this.sheet.cssRules = [];
    } else {
      // Need to reinsert for client
      _this.tag.parentNode.removeChild(_this.tag);
      _this.sheet = null;
    }
    _this.injected = false;
    return;
  };

  this.rules = function () {
    return _this.sheet.cssRules;
  };
}

var createStyleTag = function createStyleTag() {
  var tag = document.createElement('style');
  document.head.appendChild(tag);
  return tag;
};

exports.default = Stylesheet;
